File: leecode/code/238/238.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        // dp[i] = dp[0 - i-1] * dp[i+1, ...]
        int[] mul1 = new int[nums.length];
        int[] mul2 = new int[nums.length];
        mul1[0] = 1;
        mul2[nums.length - 1] = 1;


        for (int i = 1; i < nums.length; i++) {
            mul1[i] = mul1[i - 1] * nums[i - 1];
        }
        for (int i = nums.length - 2; i >= 0; i--) {
            mul2[i] = mul2[i + 1] * nums[i + 1];
        }
        for (int i = 0; i < nums.length; i++) {
            mul1[i] *= mul2[i];
        }
        return mul1;
    }
}

File: leecode/code/238/238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        rec = [1] * n
        rec2 = [1] * n
        temp = 1


        for i in range(1, n):
            rec[i] = nums[i - 1] * rec[i - 1]
        
        rec[n - 1] *= temp
        for i in range(n - 2, -1, -1):
            temp = nums[i + 1] * temp
            rec[i] *= temp


        return rec